### PR TITLE
129 Add docs, improve build, and CI

### LINKED ISSUE #129: Add better documentation for checkpoint
- [x] Higher-level introduction docs for how checkpoint serializers work
- [x] Documentation on how virtual serialize works
- [x] Checkpointing vs serialize APIs

@uhetmaniuk Can you create a docs directly here for the markdown? We can add to VT documentation generation the markdown directory so it gets generated. You can take a look in VT in how it's generated with the docs .github workflow. It might also be nice to be able to generate the oxygen separately (not sure if it's work setting that up or not).

@cz4rs Please add comments here for other documentation you would like added.

### PR DESCRIPTION
**Pull request by uhetmaniuk**
Closes #129

### CODE REVIEW

File: .github/workflows/build-docs.yml
Comment (cz4rs) [inline .github/workflows/build-docs.yml]: branch name in the comment does not match actual value (master vs develop) and the comment itself could probably be skipped completely

File: docker-compose.yml
Comment (cz4rs) [inline docker-compose.yml:50]: do we need as many volumes here?

File: ci/build_cpp.sh
Comment (cz4rs) [inline ci/build_cpp.sh:1]: if possible, this should be used in other Dockerfiles as well (`Dockerfile` and `Dockerfile-ubuntu`) to reduce duplication

File: docker-compose.yml
Comment (cz4rs) [inline docker-compose.yml:8]: no `.env` file added

File: docs/md/checkpoint_learn_build.md
Comment (cz4rs) [inline docs/md/checkpoint_learn_build.md]: random indent

File: examples/checkpoint_example_2.cc
Comment (cz4rs) [inline examples/checkpoint_example_2.cc]: it could be better to refer to serialization mode directly (`Packing`, `Unpacking` as defined in `base_serializer.h`) instead of `s.is...()` status

File: docker-compose.yml
Comment (uhetmaniuk) [reply docker-compose.yml:50]: I do not know much about Docker. 
This file is extracted from `vt/docker-compose.yml`
Maybe @lifflander could comment which volumes are useful.

File: docker-compose.yml
Comment (uhetmaniuk) [reply docker-compose.yml:8]: Done.

File: .github/workflows/build-docs.yml
Comment (uhetmaniuk) [reply .github/workflows/build-docs.yml]: Cleaned up.

File: ci/build_cpp.sh
Comment (uhetmaniuk) [reply ci/build_cpp.sh:1]: I do not understand your comment. Could you please expand?

File: docs/md/checkpoint_learn_build.md
Comment (uhetmaniuk) [reply docs/md/checkpoint_learn_build.md]: Fixed indent (but still keeping the sublist).

File: examples/checkpoint_example_2.cc
Comment (uhetmaniuk) [reply examples/checkpoint_example_2.cc]: Fixed.

File: ci/build_cpp.sh
Comment (cz4rs) [reply ci/build_cpp.sh:1]: the CI builds are performed based on the two mentioned Dockerfiles, see for example https://github.com/DARMA-tasking/checkpoint/blob/f64650d0e680aa309facfcf2629f3101a1b8399a/Dockerfile#L25-L52

`build_cpp.sh` script could probably be used there to reduce duplication (although I do understand that this is not the goal of this task, so maybe this can be solved under a separate issue after your PR gets merged)

Comment (uhetmaniuk) [reply ci/build_cpp.sh:1]: Thank you for clarifying. 
If needed, I am willing to update the branch.
But I would agree that a separate issue may be more appropriate.

**Comment by lifflander** [comment]
@uhetmaniuk The history on this PR is really screwed up. I can't cleanly rebase this and seem to be getting strange conflicts with itself. Did you try to merge or something?

Edit: Ah.. I see the merges now in the history. Please never merge under any circumstances. Always rebase.

This code would be trivial to bring up-to-date, but now it's a complex difficult process.

**Comment by lifflander** [comment]
@uhetmaniuk I completely removed your history and rewrote it cleanly in logical chunks. You should probably learn how to do this yourself. If your history is totally screwed with merges or whatever, you ought to just start over with your commits and rebuild the history from there.

```bash
git checkout -b clean_branch
git merge messy_branch
git reset --soft develop
# either commit now or unstage/restage in pieces that make sense
```

- I'm going to work on the docker tomorrow to get it where it needs to be. We don't need compose for checkpoint.
- I had a do a bunch of whitespace fixes (extra spaces)
- I don't want to remove `enum struct` because that changes the semantic, so I left that out. I'll try to figure out how to work around that bug
- I will also look at why a lot of that mcss is commented out

**Comment by uhetmaniuk** [comment]
@lifflander 

> * I will also look at why a lot of that mcss is commented out

Which comments are you referring to? 
The lines with `##!` are recognized by `m.css`

Screen Shot 2020-11-18 at 8 50 17 AM

**Comment by lifflander** [comment]
> @lifflander
> 
> > * I will also look at why a lot of that mcss is commented out
> 
> Which comments are you referring to?
> The lines with `##!` are recognized by `m.css`
> 
> Screen Shot 2020-11-18 at 8 50 17 AM

Ah, I didn't realize that. Thanks for pointing that out.

File: docker-compose.yml
Comment (lifflander) [reply docker-compose.yml:50]: Done

**Comment by lifflander** [comment]
I'm going to do a bit more work on the docs before this is done

File: ci/deps/kokkos-kernels.sh
Comment (cz4rs) [inline ci/deps/kokkos-kernels.sh]: shellcheck warns:

> SC2086: Double quote to prevent globbing and word splitting.
Did you mean:
wget http://github.com/kokkos/kokkos-kernels/archive/"${kokkos_zip_name}"

```suggestion
wget http://github.com/kokkos/kokkos-kernels/archive/"${kokkos_zip_name}"
```

File: ci/deps/kokkos-kernels.sh
Comment (cz4rs) [inline ci/deps/kokkos-kernels.sh]:
```suggestion
unzip "${kokkos_zip_name}"
```

File: ci/deps/kokkos-kernels.sh
Comment (cz4rs) [inline ci/deps/kokkos-kernels.sh]:
```suggestion
ls "${KOKKOS_ROOT}"
```

File: ci/deps/kokkos-kernels.sh
Comment (cz4rs) [inline ci/deps/kokkos-kernels.sh]:
```suggestion
ls "${KOKKOS_ROOT}"/cmake
```

File: ci/deps/cmake.sh
Comment (cz4rs) [inline ci/deps/cmake.sh]:
```suggestion
wget http://github.com/Kitware/CMake/releases/download/v"${cmake_version}"/"${cmake_tar_name}"
```

File: ci/deps/cmake.sh
Comment (cz4rs) [inline ci/deps/cmake.sh]:
```suggestion
tar xzf "${cmake_tar_name}" --one-top-level=cmake --strip-components 1
```

File: ci/deps/kokkos.sh
Comment (cz4rs) [inline ci/deps/kokkos.sh]:
```suggestion
wget http://github.com/kokkos/kokkos/archive/"${kokkos_zip_name}"
```

File: ci/deps/kokkos.sh
Comment (cz4rs) [inline ci/deps/kokkos.sh]:
```suggestion
unzip "${kokkos_zip_name}"
```

File: ci/deps/mpich.sh
Comment (cz4rs) [inline ci/deps/mpich.sh]:
```suggestion
wget http://www.mpich.org/static/downloads/"${mpich_version}"/"${mpich_tar_name}"
```

File: ci/deps/mpich.sh
Comment (cz4rs) [inline ci/deps/mpich.sh]:
```suggestion
tar xzf "${mpich_tar_name}"
```

File: ci/deps/mpich.sh
Comment (cz4rs) [inline ci/deps/mpich.sh]:
```suggestion
rm "${mpich_tar_name}"
```

File: ci/deps/mpich.sh
Comment (cz4rs) [inline ci/deps/mpich.sh]:
```suggestion
cd "${mpich_name}"
```

File: ci/deps/mpich.sh
Comment (cz4rs) [inline ci/deps/mpich.sh]:
```suggestion
make "${make_flags}"
```

File: ci/deps/mpich.sh
Comment (cz4rs) [inline ci/deps/mpich.sh]:
```suggestion
rm -rf "${mpich_name}"
```

File: docker-compose.yml
Comment (cz4rs) [inline docker-compose.yml:16]: this comment needs to be removed? (icc, nvcc)

File: docs/md/checkpoint_learn_build.md
Comment (cz4rs) [inline docs/md/checkpoint_learn_build.md]: is `detector` required or optional? (`HAS_DETECTION_COMPONENT` is used throughout the codebase)

File: docs/md/checkpoint_learn_build.md
Comment (cz4rs) [inline docs/md/checkpoint_learn_build.md:80]: same as in compose file, Intel and Nvidia are missing at the moment (alpine as well)

File: docs/md/mainpage.md
Comment (cz4rs) [inline docs/md/mainpage.md]:
```suggestion
user wants to send over the network, the traversal will create a byte buffer that
```

File: docs/md/mainpage.md
Comment (cz4rs) [inline docs/md/mainpage.md]: is mentioning intrusive / non-intrusive too much details here?

File: docs/shared/checkpoint_learn_serialize.md
Comment (cz4rs) [inline docs/shared/checkpoint_learn_serialize.md]:
```suggestion
*checkpoint* translates the object into a set of contiguous bits
```

File: docs/shared/checkpoint_learn_serialize.md
Comment (cz4rs) [inline docs/shared/checkpoint_learn_serialize.md]: `std::shared_ptr` supports footprinting only (it's guarded with apropriate `enable_if`)

**Review by cz4rs — COMMENTED** [review]
only minor issues so far, publishing first batch

File: docs/shared/checkpoint_learn_serialize.md
Comment (cz4rs) [inline docs/shared/checkpoint_learn_serialize.md]: no mention of tagged constructor

File: scripts/workflows.ini
Comment (cz4rs) [inline scripts/workflows.ini:48]: build configurations possibly missing:
- mpich disabled (for "general purpose" scenario?)
- debug
- warnings as errors enabled

File: CMakeLists.txt
Comment (PhilMiller) [inline CMakeLists.txt:66]: Do we really need to check that GCC is version > 4.8? We're most certainly using C++ features that would fail on older GCC anyway. I think the language flags setting elsewhere should fail to configure with a too-old compiler.

File: tests/unit/test_commons.h
Comment (cz4rs) [inline tests/unit/test_commons.h]: is this formatting for doc or accidental?

File: ci/deps/kokkos-kernels.sh
Comment (PhilMiller) [reply ci/deps/kokkos-kernels.sh]: It would probably be preferable to just quote the entire argument, in most of these cases, rather than just the variable expansions.

File: ci/docker/ubuntu-20.04-clang-cpp.dockerfile
Comment (PhilMiller) [inline ci/docker/ubuntu-20.04-clang-cpp.dockerfile:1]: This seems not to be what it says on the label - are we working with 20.04, or 18.04?

File: docs/md/checkpoint_learn_build.md
Comment (PhilMiller) [inline docs/md/checkpoint_learn_build.md]: Indentation mismatch.

File: tests/unit/test_commons.h
Comment (cz4rs) [inline tests/unit/test_commons.h:158]: warning spotted in this file (not related to this PR):
```
/checkpoint/tests/unit/test_commons.h:180:28: warning: moving a local object in a return statement prevents copy elision [-Wpessimizing-move]
/checkpoint/tests/unit/test_commons.h:180:28: note: remove 'std::move' call
```

File: docs/md/mainpage.md
Comment (PhilMiller) [reply docs/md/mainpage.md]: I would suggest phrasing this as "... any C++ type that has a serialize method or free function associated with it . . ."

**Review by cz4rs — APPROVED** [review]
looks good overall, random improvements suggested

File: docs/shared/checkpoint_learn_serialize.md
Comment (PhilMiller) [inline docs/shared/checkpoint_learn_serialize.md]: Please use consistent terminology - "set of bits" vs "series of bytes". I prefer the latter.

File: docs/shared/checkpoint_learn_serialize.md
Comment (PhilMiller) [inline docs/shared/checkpoint_learn_serialize.md]: Please refer to "C++ standard library" rather than "STL"

File: docs/shared/checkpoint_learn_serialize.md
Comment (PhilMiller) [inline docs/shared/checkpoint_learn_serialize.md]: Is `std::tuple` supported? If not, it should be.

Comment (cz4rs) [reply docs/shared/checkpoint_learn_serialize.md]: `checkpoint/container/tuple_serialize.h` looks legit, it should be added to the list

File: docs/shared/checkpoint_learn_serialize.md
Comment (PhilMiller) [inline docs/shared/checkpoint_learn_serialize.md]: For the documentation, I think we should strive to use language-preferred terminology - a "polymorphic class"

File: docs/shared/checkpoint_learn_serialize.md
Comment (PhilMiller) [inline docs/shared/checkpoint_learn_serialize.md]: There's nothing in the usage of these mechanisms that is specific to abstract classes and pure virtual methods. The root of the hierarchy can define every one of its methods, and this will all be as functional and desirable.

So, 

> Consider the situation where the class `BaseVC` is polymorphic (i.e. is has one or more virtual methods) and the class `CI` is derived from `BaseVC`

File: examples/checkpoint_example_virtual_macro.cc
Comment (PhilMiller) [inline examples/checkpoint_example_virtual_macro.cc]: This is deprecated. Please use the no-argument `checkpoint_virtual_serialize_root()` form

File: examples/checkpoint_example_virtual_macro.cc
Comment (PhilMiller) [inline examples/checkpoint_example_virtual_macro.cc]: Ditto deprecation - `checkpoint_virtual_serialize_derived_from( MyBase )`

**Comment by PhilMiller** [comment]
This made me realize that we've been inconsistent in how we approach deserialization of classes that don't support default construction. For some, we have a static `reconstruct` member function. For others, we have a constructor `T(SERIALIZE_RECONSTRUCT_TAG)`. We should pick one or the other, and standardize on it, to be consistent for users and to ease implementation of stuff like #161.

**Comment by PhilMiller** [comment]
Looks mostly good. I just converted it to a 'Draft' so that no one would plow ahead with hitting the Merge button before everyone actually agrees it's ready.

File: ci/deps/kokkos-kernels.sh
Comment (lifflander) [reply ci/deps/kokkos-kernels.sh]: I get why these checkers throw these warnings, but I do know that the `kokkos_zip_name` never contains a space. However, I will make these changes.

File: docker-compose.yml
Comment (lifflander) [reply docker-compose.yml:16]: I'm planning on adding icc/nvcc next

File: docs/md/checkpoint_learn_build.md
Comment (lifflander) [reply docs/md/checkpoint_learn_build.md]: Good point, it is optional

File: docs/md/mainpage.md
Comment (lifflander) [reply docs/md/mainpage.md]: Fixed

File: ci/docker/ubuntu-20.04-clang-cpp.dockerfile
Comment (lifflander) [reply ci/docker/ubuntu-20.04-clang-cpp.dockerfile:1]: This is a symbolic link to 18.04. Both 18.04 and 20.04 are used for the containers due to the compilers that are available across the linux versions as packages

File: docs/shared/checkpoint_learn_serialize.md
Comment (lifflander) [reply docs/shared/checkpoint_learn_serialize.md]: Added to docs

File: docker-compose.yml
Comment (lifflander) [reply docker-compose.yml:16]: Done

File: docs/md/checkpoint_learn_build.md
Comment (lifflander) [reply docs/md/checkpoint_learn_build.md:80]: Added

File: scripts/workflows.ini
Comment (lifflander) [reply scripts/workflows.ini:48]: Added warnings as errors build. I don't think that non-MPI makes sense because it just disables the MPI examples.

File: tests/unit/test_commons.h
Comment (lifflander) [reply tests/unit/test_commons.h]: Fixed

File: tests/unit/test_commons.h
Comment (lifflander) [reply tests/unit/test_commons.h:158]: Fixed

File: examples/checkpoint_example_virtual_macro.cc
Comment (lifflander) [reply examples/checkpoint_example_virtual_macro.cc]: Fixed

File: examples/checkpoint_example_virtual_macro.cc
Comment (lifflander) [reply examples/checkpoint_example_virtual_macro.cc]: Fixed

File: docs/shared/checkpoint_learn_serialize.md
Comment (lifflander) [reply docs/shared/checkpoint_learn_serialize.md]: Removed

File: docs/md/mainpage.md
Comment (lifflander) [reply docs/md/mainpage.md]: Fixed

File: docs/md/checkpoint_learn_build.md
Comment (lifflander) [reply docs/md/checkpoint_learn_build.md]: Fixed

File: CMakeLists.txt
Comment (lifflander) [reply CMakeLists.txt:66]: Probably not, but I'll leave this in there just to be sure. I think checkpoint will compile with old compilers.

File: docs/shared/checkpoint_learn_serialize.md
Comment (lifflander) [reply docs/shared/checkpoint_learn_serialize.md]: Fixed

File: docs/shared/checkpoint_learn_serialize.md
Comment (lifflander) [reply docs/shared/checkpoint_learn_serialize.md]: Done

File: docs/shared/checkpoint_learn_serialize.md
Comment (lifflander) [reply docs/shared/checkpoint_learn_serialize.md]: Fixed

File: ci/deps/kokkos-kernels.sh
Comment (lifflander) [reply ci/deps/kokkos-kernels.sh]: Fixed

File: docs/shared/checkpoint_learn_serialize.md
Comment (lifflander) [reply docs/shared/checkpoint_learn_serialize.md]: Fixed

**Review by cz4rs — APPROVED** [review]
looks great!

**Pull request merged**
